tools/src/main/java/org/apache/kafka/tools/MessageReader.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.tools;
+
+import org.apache.kafka.clients.producer.ProducerRecord;
+
+import java.io.InputStream;
+import java.util.Properties;
+
+/**
+ * Typical implementations of this interface convert data from an {@link InputStream} received via
+ * {@link MessageReader#init(InputStream, Properties)} into a {@link ProducerRecord} instance on each
+ * invocation of `{@link MessageReader#readMessage()}`.
+ *
+ * This is used by the {@link ConsoleProducer}.
+ */
+public interface MessageReader {
```

Comment:
Note that this class is only used by the `ConsoleProducer` and has one implementation. If no further used is intended, it can be removed.

Reply:
This would be a breaking change, because there is an option that allows to provide your own message reader implementation.

Reply:
Apologies, that is right. Thanks for correcting. On that note, this interface moved from `kafka.common.MessageReader` to `org.apache.kafka.tools.MessageReader`. Any foreign implementation will have to be updated accordingly. What is the convention to follow is in this case - preserve the same package to avoid breaking dependencies or change the package to match the new structure?

Reply:
This is tricky. It's not a class from the public API, but it's exposed via the command line and definitely used (search for AvroMessageReader for example). I also found the still open KIP-641 which deals with this interface. 

We should have a deprecation period, but at the same time we don't want to depend on core. Given that checkstyle can only handle a single root package, the only solution I see is to create a :tools:deprecated sub-module containing kafka.common.MessageReader.

@mimaison @showuon WDYT?

Reply:
Agreed on taking an intermediate step to preserve FQN compatibility for the SPI.

Reply:
Could we update checkstyle to have an exception?

Reply:
If we want to eventually change the package name, we would need a KIP.

Overall, I feel like that are a few compatibility issues with the migration of those tools so a small KIP may be good anyway to discuss how we want to address all of them together instead of discussing them in every PR. What do you think?

Reply:
Thanks David for reviewing the PR. Agreed that a KIP is appropriate given the compatibility changes to SPIs you pointed out.

Reply:
Found the following impacted readers:

- [ProtobufMessageReader](https://github.com/confluentinc/schema-registry/blob/master/protobuf-serializer/src/main/java/io/confluent/kafka/formatter/protobuf/ProtobufMessageReader.java)
- [AvroMessageReader](https://github.com/confluentinc/schema-registry/blob/master/avro-serializer/src/main/java/io/confluent/kafka/formatter/AvroMessageReader.java)
- [JsonSchemaMessageReader](https://github.com/confluentinc/schema-registry/blob/master/json-schema-serializer/src/main/java/io/confluent/kafka/formatter/json/JsonSchemaMessageReader.java)

with for parent class [SchemaMessageReader](https://github.com/confluentinc/schema-registry/blob/master/schema-serializer/src/main/java/io/confluent/kafka/formatter/SchemaMessageReader.java).

A minor reference in the documentation for the [connect-streams-pipeline](https://github.com/confluentinc/examples/blob/7.3.1-post/connect-streams-pipeline/docs/index.rst).

Reply:
@Hangleton Do you already have a strategy to deal with this? How are you planning to provide a deprecation period? Can we generalize this approach?

Reply:
- For `MessageReader`: maybe we could provide a depreciation period until Apache Kafka 4.0 by exposing the class defined in [KIP-641](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=158866569) and keep the interface `kafka.common.MessageReader` inheriting from `org.apache.kafka.common.MessageReader` to avoid breaking the implementations above? The class `org.apache.kafka.common.MessageReader` could be put in the `clients` module (KIP-641 suggests to have `MessageReader` in this module since it is a public interface).
- For the JMX Tool: maybe we could keep the existing class with a deprecation warning as you did in [PR-13195](https://github.com/apache/kafka/pull/13195) and create a shell and bat scripts as the public contract for the new class?
- `kafka.tools.StateChangeLogMerger` could be removed if it isn't used as mentioned in [PR-13171](https://github.com/apache/kafka/pull/13171).
- There is a missing BAT script for `FeatureCommand` although the shell script is defined. This could be added. 
- I skimmed through the other command/tool classes and all of them seem to be exposed via a script so that there shouldn't be any compatibility problem there? What other changes do we want to address?

Reply:
Thanks. This is more or less what I had in mind. I'm going to go through all tools to see if there is any further issue and then we can discuss the policy with the community and possibly have a vote.

Reply:
https://cwiki.apache.org/confluence/display/KAFKA/KIP-906%3A+Tools+migration+guidelines

We also revived KIP-641. Any feedback is welcomed.

Reply:
Thanks @fvaleri and apologies for the delay, I will look at the KIP. Thanks!